File: spring/spring_tutorial/my-app/src/main/java/com/karma/myapp/config/SecurityConfig.java
package com.karma.myapp.config;

import com.karma.myapp.exception.CustomAuthenticationEntryPoint;
import com.karma.myapp.service.UserAccountService;
import com.karma.myapp.utils.JwtFilter;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.security.servlet.PathRequest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;

import java.util.Arrays;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {

    private final UserAccountService userAccountService;
    @Value("${jwt.secret-key}") private String secretKey;

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        return http
                // csrf
                .csrf()
                .disable()                // cors : 모두 허용
                .cors(c -> {
                            CorsConfigurationSource source = request -> {
                                CorsConfiguration config = new CorsConfiguration();
                                config.setAllowedOrigins(
                                        Arrays.asList("http://localhost:3000")
                                );
                                config.setAllowedMethods(
                                        Arrays.asList("GET", "POST", "PUT", "DELETE")
                                );
                                config.setAllowedHeaders(Arrays.asList("Content-Type", "Authorization"));
                                return config;
                            };
                            c.configurationSource(source);
                        }
                )
                // session - JWT를 사용하므로 session은 사용하지 않으므로, Stateless
                .sessionManagement()
                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                .and()
                .authorizeHttpRequests(
                        auth -> auth
                                // static files(html,css,js,favicon...) → permit all
                                .requestMatchers(PathRequest.toStaticResources().atCommonLocations()).permitAll()
                                // articles → GET request permit all
                                .requestMatchers(
                                        HttpMethod.GET,
                                        "/api/article", "api/article/*", "api/comment", "api/emotion"
                                ).permitAll()
                                // sign up, login → POST request permit all
                                .requestMatchers(
                                        HttpMethod.POST,
                                        "/api/user/signup", "api/user/login"
                                ).permitAll()
                                // others  → authenticated
                                .anyRequest().authenticated()
                )
                // exception handling
                .exceptionHandling()
                .authenticationEntryPoint(new CustomAuthenticationEntryPoint())
                .and()
                // JWT filter
                .addFilterBefore(new JwtFilter(userAccountService, secretKey), UsernamePasswordAuthenticationFilter.class)
                .build();
    }
}

File: backend/src/main/kotlin/com/karma/chat/config/security/SecurityConfig.kt
package com.karma.chat.config.security

import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
import org.springframework.boot.autoconfigure.security.ConditionalOnDefaultWebSecurity
import org.springframework.boot.autoconfigure.security.servlet.PathRequest
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter

@Configuration
@EnableWebSecurity
@ConditionalOnDefaultWebSecurity
@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
class SecurityConfig(
    private val jwtAuthenticationFilter: JwtAuthenticationFilter
) {

    // 인증권한 없이 요청을 날릴 수 있는 end point
    private val permitUrls = arrayOf("/swagger-ui/**", "/api/auth/signup/**", "/api/auth/signin/**")

    @Bean
    fun filterChain(http: HttpSecurity) = http
        .csrf { it.disable() }
        .headers { it.frameOptions { frameOptions -> frameOptions.sameOrigin() } }
        .authorizeHttpRequests {
            it.requestMatchers(*permitUrls).permitAll()
                .anyRequest().authenticated()
        }
        // session대신 jwt를 사용하기 때문에, stateless로 설정
        .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
        // jwt 인증필터 → 기본 인증필터 순으로 chaining
        .addFilterBefore(jwtAuthenticationFilter, BasicAuthenticationFilter::class.java)
        .build()!!

    @Bean
    fun passwordEncoder() = BCryptPasswordEncoder()
}
